### PR TITLE
GetSymbolInfo and GetTypeInto for new pattern forms

### PR DESCRIPTION
**Pull request by gafter**

### CODE REVIEW

File: src/Compilers/CSharp/Portable/BoundTree/BoundNodes.xml [hunk 1]
```diff
@@ -1949,6 +1949,7 @@
 
   <AbstractNode Name="BoundPattern" Base="BoundNode">
     <Field Name="InputType" Type="TypeSymbol" Null="disallow"/>
+    <Field Name="ConvertedType" Type="TypeSymbol" Null="disallow"/>
```
Comment (alrz) [inline src/Compilers/CSharp/Portable/BoundTree/BoundNodes.xml]: Isn't this the same as *output type*? If so we could probably unify.

Comment (gafter) [reply src/Compilers/CSharp/Portable/BoundTree/BoundNodes.xml]: We will unify when the two branches merge.  The spec doesn't use either term.

File: src/Compilers/CSharp/Test/Semantic/Semantics/PatternMatchingTests3.cs [hunk 2]
```diff
@@ -3107,5 +3103,233 @@ .locals init (char V_0)
...
+            for (int i = 0; i < 31; i++)
+            {
+                var pattern = patterns[i];
+                AssertEmpty(model.GetSymbolInfo(pattern));
+                switch (i)
+                {
+                    case 0:
+                        Assert.Equal("((N, N))", pattern.ToString());
+                        Assert.Equal(SyntaxKind.ParenthesizedPattern, pattern.Kind());
+                        Assert.Equal("System.Object", model.GetTypeInfo(pattern).Type.ToTestDisplayString());
```
Comment (333fred) [inline src/Compilers/CSharp/Test/Semantic/Semantics/PatternMatchingTests3.cs]: I'll admit this surprises me. Why is it object?

Comment (333fred) [reply src/Compilers/CSharp/Test/Semantic/Semantics/PatternMatchingTests3.cs]: And I guess that follows for the rest of the `Object` assertions below.

Comment (gafter) [reply src/Compilers/CSharp/Test/Semantic/Semantics/PatternMatchingTests3.cs]: In this way patterns are the exact opposite of expressions.

In an expression, types flow from the inside to the outside, possibly getting converted as they go.

In a pattern, types flow from the outside to the inside, possibly getting converted as they go.

In this particular case, the expression coming in is of type `object`.  However, it is converted to the type `ITuple` by the pattern in order to be matched.

---
In reply to: [391898883](https://github.com/dotnet/roslyn/pull/42357#discussion_r391898883) [](ancestors = 391898883)

File: src/Compilers/CSharp/Portable/Binder/Binder_Operators.cs [hunk 3]
```diff
@@ -2794,7 +2794,9 @@ private BoundExpression BindIsOperator(BinaryExpressionSyntax node, DiagnosticBa
                     var hasErrors = node.Right.HasErrors || constantValueOpt is null;
                     isTypeDiagnostics.Free();
                     diagnostics.AddRangeAndFree(isPatternDiagnostics);
-                    var boundConstantPattern = new BoundConstantPattern(node.Right, convertedExpression, constantValueOpt ?? ConstantValue.Bad, operand.Type, hasErrors) { WasCompilerGenerated = true };
+                    var boundConstantPattern = new BoundConstantPattern(
+                        node.Right, convertedExpression, constantValueOpt ?? ConstantValue.Bad, operand.Type, convertedExpression.Type ?? operand.Type, hasErrors)
+                    { WasCompilerGenerated = true };
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Binder/Binder_Operators.cs]: >{ WasCompilerGenerated = true }; [](start = 20, length = 32)

Consider shifting initializer to the right, at least on one tab.

Comment (gafter) [reply src/Compilers/CSharp/Portable/Binder/Binder_Operators.cs]: Will do, and will suppress the resulting warning about whitespace.

File: src/Compilers/CSharp/Test/Semantic/Semantics/PatternMatchingTests3.cs [hunk 4]
```diff
@@ -3107,5 +3103,467 @@ .locals init (char V_0)
...
+                        Assert.Equal("System.Char c9", model.GetSymbolInfo(((RelationalPatternSyntax)pattern).Expression).Symbol.ToTestDisplayString());
+                        Assert.Equal("System.Object", model.GetTypeInfo(pattern).Type.ToTestDisplayString());
+                        Assert.Equal("System.Char", model.GetTypeInfo(pattern).ConvertedType.ToTestDisplayString());
+                        break;
+                }
+            }
+        }
+
+
+        [Fact]
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Semantic/Semantics/PatternMatchingTests3.cs]: >[Fact] [](start = 8, length = 6)

It looks like there is an extra empty line above (I see two).

File: src/Compilers/CSharp/Test/Semantic/Semantics/PatternMatchingTests3.cs [hunk 5]
```diff
@@ -3107,5 +3103,467 @@ .locals init (char V_0)
...
+            // Conjunctive and disjunctive patterns
+            case (>= A and <= Z) or (>= a and <= z): break;
+            // Negated patterns
+            case not (> c0 and < c9): break;
+        }
+    }
+}
+";
+            var compilation = CreatePatternCompilation(source, options: TestOptions.DebugDll);
+            compilation.VerifyDiagnostics(
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Semantic/Semantics/PatternMatchingTests3.cs]: >( [](start = 41, length = 1)

Consider removing the line break if expect no diagnostics. In other tests too.

Comment (gafter) [reply src/Compilers/CSharp/Test/Semantic/Semantics/PatternMatchingTests3.cs]: This is a convention used throughout this file, and I prefer it to emphasize that the diagnostics are empty, and to make it easy to change if that changes.

---
In reply to: [392351108](https://github.com/dotnet/roslyn/pull/42357#discussion_r392351108) [](ancestors = 392351108)

File: src/Compilers/CSharp/Test/Semantic/Semantics/PatternMatchingTests3.cs [hunk 6]
```diff
@@ -3107,5 +3103,466 @@ .locals init (char V_0)
...
+            var tree = compilation.SyntaxTrees[0];
+            var model = compilation.GetSemanticModel(tree);
+
+            var patterns = tree.GetRoot().DescendantNodes().OfType<PatternSyntax>().ToArray();
+            Assert.Equal(31, patterns.Length);
+            for (int i = 0; i < 31; i++)
+            {
+                var pattern = patterns[i];
+                AssertEmpty(model.GetSymbolInfo(pattern));
+                switch (i)
```
Comment (agocke) [inline src/Compilers/CSharp/Test/Semantic/Semantics/PatternMatchingTests3.cs]: To improve readability, consider switching on `pattern.ToString()`

**Pull request merged**